source/lac/sparsity_tools.cc

```diff
@@ -39,15 +39,275 @@ extern "C"
+
+      // one more nuisance: we have to copy our own data to arrays that store
+      // signed integers :-(
+      std::vector<idx_t> int_rowstart(1);
+      int_rowstart.reserve(sparsity_pattern.n_rows()+1);
+      std::vector<idx_t> int_colnums;
+      int_colnums.reserve(sparsity_pattern.n_nonzero_elements());
+      for (SparsityPattern::size_type row=0; row<sparsity_pattern.n_rows(); ++row)
+        {
+          for (SparsityPattern::iterator col=sparsity_pattern.begin(row);
+               col < sparsity_pattern.end(row); ++col)
+            int_colnums.push_back(col->column());
+          int_rowstart.push_back(int_colnums.size());
+        }
+
+      std::vector<idx_t> int_partition_indices (sparsity_pattern.n_rows());
+
+      // Make use of METIS' error code.
+      int ierr;
+
+      // Select which type of partitioning to create
+
+      // Use recursive if the number of partitions is less than or equal to 8
+      if (nparts <= 8)
+        ierr = METIS_PartGraphRecursive(&n, &ncon, int_rowstart.data(), int_colnums.data(),
+                                        nullptr, nullptr, nullptr,
+                                        &nparts,nullptr,nullptr,options,
+                                        &dummy,int_partition_indices.data());
+
+      // Otherwise use kway
+      else
+        ierr = METIS_PartGraphKway(&n, &ncon, int_rowstart.data(), int_colnums.data(),
+                                   nullptr, nullptr, nullptr,
+                                   &nparts,nullptr,nullptr,options,
+                                   &dummy,int_partition_indices.data());
+
+      // If metis returns normally, an error code METIS_OK=1 is returned from
+      // the above functions (see metish.h)
+      AssertThrow (ierr == 1, ExcMETISError (ierr));
+
+      // now copy back generated indices into the output array
+      std::copy (int_partition_indices.begin(),
+                 int_partition_indices.end(),
+                 partition_indices.begin());
+#endif
+    }
+
+//Query function unused if zoltan is not installed
+#ifdef DEAL_II_TRILINOS_WITH_ZOLTAN
+    //Query functions for partition_zoltan
+    int get_number_of_objects(void *data, int *ierr)
+    {
+      SparsityPattern *graph = reinterpret_cast<SparsityPattern *>(data);
+
+      *ierr = ZOLTAN_OK;
+
+      return graph->n_rows();
+    }
+#endif
+
+//Query function unused if zoltan is not installed
+#ifdef DEAL_II_TRILINOS_WITH_ZOLTAN
+    void get_object_list(void *data, int sizeGID, int sizeLID,
+                         ZOLTAN_ID_PTR globalID, ZOLTAN_ID_PTR localID,
+                         int wgt_dim, float *obj_wgts, int *ierr)
+    {
+      SparsityPattern *graph = reinterpret_cast<SparsityPattern *>(data);
+      *ierr = ZOLTAN_OK;
+
+      Assert( globalID != nullptr , ExcInternalError() );
+      Assert( localID != nullptr , ExcInternalError() );
+
+      //set global degrees of freedom
+      auto n_dofs = graph->n_rows();
+      for (int i=0; i < n_dofs; i++)
```

Comment:
in order to avoid warnigns
```
/home/bob/source/source/lac/sparsity_tools.cc:165:23: warning: comparison of integers of different signs: 'int' and 'unsigned int' [-Wsign-compare]
      for (int i=0; i < n_dofs; i++)
```

 do: `for (unsigned int i = 0; ....`